### ISSUE TITLE
Adding domain specific vocabulary

### ISSUE THREAD

**Issue by artemisart** [google-research/bert#9]
Hi, thanks for the release !

I will need to add some domain specific vocabulary, do you have any suggestion on how to do it ?
I was thinking of replacing some [unused#] tokens in the vocab file (so if i'm not mistaken they already have existing weights in the checkpoint model files) to avoid extending the matrices, and then finetuning the LM with a domain specific corpus.
If it feasible I would also try to do a first LM finetuning pass with the existing vocab embeddings freezed, to only learn the new words, and then a second pass with everything unfreezed.

Do you think it's the right way to do it ?
How can I freeze a subset of the embeddings ? (Gradient masking ?)

**Comment by alznn** [comment]
@jacobdevlin-google 
Thank you for sharing!
However, I am still a little bit confuse about the step you mention below  **"generates a new checkpoint that is identical to the pre-trained checkpoint"** .

First , is this meant I should create a new script or just execute run_pretrained. py?

Second, is this meant that there will be a brand new checkpoint or they're still an original checkpoint (download from Readme) but trained with new vocb size?

sorry these might be stupid questions but thanks in advance

**Comment by stevenwernercs** [comment]
Can you point out how to 'run pre-trianing'?  (specific *.py script or function)
The use of 'pre-trained' is overloaded and effects google results.

I want to take an existing 'pre-trained' BERT model, and then extend/append it by 'pre-training' it with unlabeled unsupervised text in my specific domain.

I know how to fine-tune it on a downstream task, but am unsure how to continue the embedding layer training beyond what is provided in the 'pre-trained' models pre-trained on Wikipedia

Thank you

**Comment by gro1m** [comment]
Hi @jacobdevlin-google 
I can add the vocab, but actually I would like to pass sentences as input and not just vocabulary words, because I would like to profit from the masked language modeling scheme that BERT uses. It is unclear to me how that should work when only passing the vocabulary words but not the sentences. Can you enlighten me here or give me a suggestion on how to finetune with sentences as input?

**Comment by kumarme072** [comment]
@jacobdevlin-google 
I have come across many similar issues asking about how to add new tokens to a vocabulary, for reference, here are a couple links to useful comments made for doing roughly that:

https://github.com/huggingface/transformers/issues/1413#issuecomment-538083512
https://github.com/huggingface/transformers/issues/2691#issuecomment-587473545
https://github.com/huggingface/tokenizers/issues/627#issuecomment-784286485
However, I am concerned with how to first identify tokens that make sense to add to an existing tokenizer's vocabulary, and also possibly whether or not it makes sense to consider removing tokens from a vocabulary.

Some context into my situation:

My situation I believe is quite typical: I have a reasonably large domain-specific text dataset available, all in English, and my end goal is produce a domain-specific language model that can be used for various downstream NLP tasks, (e.g., text classification, sentence similarity, etc), after additional fine-tuning on those tasks.

But from my current understanding, to first obtain that domain-specific language model, I basically have two options:

train a tokenizer from scratch and then use that tokenizer to train a LM from scratch.
modify the vocabulary of a pretrained tokenizer, adjust a (also pretrained) LM's embedding matrix to work with this new vocab size, and fine-tune the pretrained LM on my domain-specific text dataset on something like MLM.
I am struggling with the first option because (as far as I know) training a language model from scratch is quite expensive, and although I do have some budget for this, I do not have on the order of thousands of dollars.

I am starting to explore the second option, but I am confused on how to properly modify the vocabulary in a way that makes sense for the model, and concerned about what other side effects this could cause.

To summarize:

I'd really like to know if there is a low cost option for training a LM from scratch to do option 1 above
Or, if option 2 makes more sense, how to properly modify a vocabulary (find good new tokens, remove unused ones, etc), and adapt the model to overcome potential negative side effects of messing with the embeddings.
Thanks for the help. Sorry for a long question but I thought some context may be needed since I might be asking the wrong question in the first place. Cheers.  ##Someone already asked this but no satisfactory answer was there....
